Minor change to http:options and path-combine

owaddell opened the pull request:
This pull request
- adds `validate-path` to `http:options` and exposes `http:valid-path?` so a server can supply a `http:url-handler` that interprets paths that would otherwise be invalid, yet still check the extracted paths before hitting the filesystem, and
- modifies `path-combine` to avoid adding extra directory separators while allocating less in some cases.
- [edit] fixes a bug in `http:eval-dynamic-page`

owaddell-beckman committed:
add validate-path to http:options

Add validate-path to http:options and expose http:valid-path? so
a server can supply a http:url-handler that interprets paths that
would otherwise be invalid, yet still check the extracted paths
before hitting the filesystem.

owaddell-beckman committed:
avoid extra directory-separator in path-combine

path-combine now preserves directory separators present in its arguments,
but avoids adding extra directory separators, e.g., when a constituent
begins with a directory separator. It may also allocate less.

owaddell-beckman committed:
fix bug in http:eval-dynamic-page

burgerrg commented on doc/swish/http.tex:
```diff
@@ -303,7 +303,12 @@ \subsection{WebSocket Protocol}\label{sec:websocket-protocol}
+
+  \code{validate-path} & \code{http:valid-path?} & a predicate;
+  the server raises an exception if the \code{<request>} \var{path}
+  does not satisfy this prediate. \\
```
prediate => predicate

owaddell replied:
Oh dear, thanks for finding this. Fixed.

burgerrg commented on doc/swish/http.tex:
```diff
@@ -682,6 +687,16 @@ \subsection{WebSocket Protocol}\label{sec:websocket-protocol}
 
+\defineentry{http:valid-path?}
+\begin{procedure}
+  \code{(http:vavlid-path? \var{path})}
```
vavlid => valid

owaddell replied:
Thanks for finding this. Fixed.

laqrix reviewed (approved):
Nice work Oscar!

owaddell-beckman committed:
ci: scale another timeout and improve test:receive

Scale timeout in response to another sporadic GitHub Actions macOS test failure.

Update test:receive
 - fix inadvertent reference to timeout pattern variable in 'timeout
 - report source for individual clause rather than the entire expression

owaddell commented:
Thanks, guys.

I noticed a failed macOS test and pushed an update to use `test:receive` in that mat.

owaddell-beckman committed:
fix broken http redirect for /

The recent change to path-combine means that (path-combine "abc" "")
now returns "abc" instead of "abc/". It looks like the http server had
relied on that and may also have been using substring to work around the
case where (path-combine "a" "/b") used to return "a//b".
The simplification here fixes redirect for /.

owaddell commented:
Oops. One more wafer-thin commit.

owaddell commented:
Thanks for reviewing this, gentlemen.
